src/proxy/ILock.ts

```diff
@@ -0,0 +1,73 @@
+export interface Lock {
+
+    /**
+     * Acquires the lock, waiting indefinitely if necessary.
+     * The returned promise will be resolved as soon as the lock becomes available.
+     * After this method is called the user must call `unlock` to release the lock.
+     */
+    lock(): Promise<void>;
+
+    /**
+     * Leases the lock for a specified period of time, waiting indefinitely if necessary.
+     * The returned promise will be resolved as soon as the lock becomes available.
+     * The lock will be held for the specified amount of time and then released automatically.
+     *
+     * @param leaseMillis period of time in milliseconds for which the lock should be held.
+     */
+    lease(leaseMillis: number): Promise<void>;
+
+    /**
+     * Tries to acquire the lock, but does not wait for it to become available.
+     * @returns `true` if lock was obtained right away, `false` otherwise.
+     */
+    tryLock(): Promise<boolean>;
+
+    /**
+     * Tries to acquire the lock, waiting for the specified amount of time.
+     * @param waitMillis period of time in milliseconds to wait for the lock to become available.
+     * @returns `true` if the lock was obtained within the specified period, `false` otherwise.
+     */
+    waitLock(waitMillis: number): Promise<boolean>;
```

Comment:
I think this is "with lease" version of `tryLock`. This can be an optional argument in `tryLock`.